UI/src/view/AssetLabel.js

```diff
@@ -33,23 +35,69 @@
         };
 
         this.renderFeatures = function(assets, zoomLevel, getPoint){
-            if(!me.isVisibleZoom(zoomLevel))
-                return [];
-
-            return _.chain(assets).
-                map(function(asset){
-                    var assetValue = me.getValue(asset);
-                    if(assetValue !== undefined){
-                        var style = me.getStyle(assetValue);
-                        var feature = me.createFeature(getPoint(asset));
-                        feature.setStyle(style);
-                        return feature;
-                    }
-                }).
-                filter(function(feature){ return feature !== undefined; }).
-                value();
+          if(!me.isVisibleZoom(zoomLevel))
+            return [];
+
+          return _.chain(assets).
+          map(function(asset){
+            var assetValue = me.getValue(asset);
+            if(assetValue !== undefined){
+              var style = me.getStyle(assetValue);
+              var feature = me.createFeature(getPoint(asset));
+              feature.setProperties(asset);
+              feature.setStyle(style);
+              return feature;
+            }
+          }).
+          filter(function(feature){ return feature !== undefined; })
+            .value();
         };
 
+      this.getCoordinateForGrouping = function (point){
+        var assetCoordinate = {lon : point.lon, lat : point.lat};
+        var assetCounter = {counter: 1};
+        if(_.isEmpty(me.populatedPoints)){
+          me.populatedPoints.push({coordinate: assetCoordinate, counter: 1});
+        }else{
+          var populatedPoint = _.find(me.populatedPoints, function (p) {
+            return me.isInProximity(point, p, me.MIN_DISTANCE);
+          });
+          if (!_.isUndefined(populatedPoint)) {
+            assetCoordinate = populatedPoint.coordinate;
+            assetCounter.counter = populatedPoint.counter + 1;
+            populatedPoint.counter++;
+          } else {
+            me.populatedPoints.push({coordinate: assetCoordinate, counter: 1});
+          }
+        }
+        return [[assetCoordinate.lon, assetCoordinate.lat], assetCounter.counter];
+      };
+
+      this.renderGroupedFeatures = function(assets, zoomLevel, getPoint){
+        if(!this.isVisibleZoom(zoomLevel))
+          return [];
+
+        return _.chain(assets).
+        map(function(asset){
+          var value = me.getValue(asset);
+          var assetLocation = getPoint(asset);
```

Comment:
Instead of using getCoordinateForGrouping, can't we use the index of the map: 

` map(function(asset, index){`

And for getting the coordinates, we can get them from the asset. 

For checking the assets that are inProximity can't we before of iterate over the assets just group them by coordinates and proximity and after that create the features.